### PR TITLE
Dependency checks

### LINKED ISSUE #89: Code quality: inter-module automatic dependency checks
Feature #32
Upstream https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/issues/827

### Description

As an EDC contributor I would like to know, if my code is creating forbidden dependencies between modules so that I can prevent them. See upstream EDC story for sample use cases.

Use approach from spike #160 .

Remaining work to be done are small cleanups from what has been done in #160.

### Acceptance Criteria
- [x] A mechanism for detecting forbidden module dependencies is in place
- [x] Forbidden module dependencies are checked during CI
- [x] It is configurable whether forbidden module dependencies will be logged or whether they fail the build

### Tasks
- [x] Introduce selected mechanism in EDC repo based on #79 
- [x] Clean up custom gradle task by extracting it from main build file (see https://docs.gradle.org/current/userguide/custom_tasks.html#sec:packaging_a_task_class)
- [ ] Unit test custom gradle task

### PR DESCRIPTION
**Pull request by marcgs**
## What this PR changes/adds

 The CI workflow is extended with two Gradle tasks for analyzing EDC module dependencies:

 - The [Dependency Analysis Gradle Plugin](https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin) is used for detecting unused dependencies and dependencies declared on the wrong configuration (`api` vs `implementation` vs `compileOnly`, etc.).
 - A custom gradle plugin defines rules on which module dependencies are allowed within EDC (e.g. no module may depend directly on any of the core modules).

 The tasks are initially set to only emit warnings, but will be configured to fail builds once all warnings have been resolved.

 The tasks are run in the CI workflow.

Outputs can be seen in two tasks in the `Test Code (Style, Tests)` workflow -> `Dependency-analysis`:
- Dependency rules report (450 violations, see "Dependency Rules" in [CI run](https://github.com/agera-edc/DataSpaceConnector/runs/5689505637?check_suite_focus=true))
- Dependency analysis report (~500 lines, see "Dependecy Analysis" in [CI run](https://github.com/agera-edc/DataSpaceConnector/runs/5689505637?check_suite_focus=true))

## Why it does that

 The [Gradle Java library plugin](https://docs.gradle.org/current/userguide/java_library_plugin.html#sec:java_library_recognizing_dependencies) defines a clear separation between configurations. This allows controlling and limiting the use of transitive dependencies.

 The Java library plugin introduces the concept of ABI (Application Binary Interface) which includes (among others) types used in public methods. Dependencies that define those types should be defined in the `api` configuration. Other dependencies should be defined in the `implementation` configuration so that they are not transitively exposed, and therefore do not leak into the consumers' compile classpath.

 The Dependency Analysis Gradle Plugin analyzes the abstract syntax tree to determine the ABI. On that basis, it can recommend which import configurations should be changed.

## Further notes

Modified some of the project files to implement the recommendations from the tool for the `core` module. This validated that the recommendations were reasonable:
* `:core:contract`:  scope of `:common:state-machine-lib` and `:common:util` changed from api to implementation. 
* `:core:contract`:  scope of `:common:state-machine-lib` changed from api to implementation. Removed unused dependency to `:common:util`

Increased compiler heap space to deal with a memory error.

## Linked Issue(s)

Closes #89

## Checklist

- [x] added appropriate tests?
- [x] performed checkstyle check locally?
- [x] added/updated copyright headers?
- [x] documented public classes/methods?
- [x] added/updated relevant documentation?
- [x] added relevant details to the changelog? (_skip with label `no-changelog`_)
- [x] formatted title correctly? (_take a look at the [CONTRIBUTING](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/CONTRIBUTING.md#submit-a-pull-request) and [styleguide](https://github.com/eclipse-dataspaceconnector/DataSpaceConnector/blob/main/styleguide.md) for details_)

### CODE REVIEW

File: extensions/api/auth-spi/src/main/java/org/eclipse/dataspaceconnector/api/auth/AuthenticationRequestFilter.java [hunk 1]
```diff
@@ -29,7 +29,7 @@ public AuthenticationRequestFilter(AuthenticationService authenticationService)
     }
 
     @Override
-    public void filter(ContainerRequestContext requestContext) {
+    public void filter(ContainerRequestContext requestContext) throws NotAuthorizedException {
```
Comment (algattik) [inline extensions/api/auth-spi/src/main/java/org/eclipse/dataspaceconnector/api/auth/AuthenticationRequestFilter.java:32]: add comment explaining why RuntimeException is explicitly declared?

and explain this in the ADR maybe?

File: build.gradle.kts [hunk 2]
```diff
@@ -232,4 +235,52 @@ openApiMerger {
...
+                        "org.mockito:mockito-core",
+                    )
+                }
+                onIncorrectConfiguration {
+                    exclude(
+                        // some common dependencies are intentionally exported by core:base for simplicity
+                        "com.squareup.okhttp3:okhttp",
+                        "net.jodah:failsafe",
+                        // public methods annotated with @WithSpan
+                        "io.opentelemetry:opentelemetry-extension-annotations",
```
Comment (algattik) [inline build.gradle.kts:277]: with rc04 should be able to use excludeAnnotations instead https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin/issues/610

**Review by algattik — CHANGES_REQUESTED** [review]
inline

File: build.gradle.kts (hunk 2 above)
Comment (marcgs) [reply build.gradle.kts]: Done.

File: extensions/api/auth-spi/src/main/java/org/eclipse/dataspaceconnector/api/auth/AuthenticationRequestFilter.java (hunk 1 above)
Comment (marcgs) [reply extensions/api/auth-spi/src/main/java/org/eclipse/dataspaceconnector/api/auth/AuthenticationRequestFilter.java:32]: Is it correct that the reason is to have `api-core` as an API dependency in `auth-spi`? Not sure how maintainable this is, I think developers are not used to declare RuntimeExceptions and there is no failing check at the moment that would verify this.

Comment (marcgs) [reply extensions/api/auth-spi/src/main/java/org/eclipse/dataspaceconnector/api/auth/AuthenticationRequestFilter.java]: Removed Exception from signature and marked it with `@ExcludeFromDependecyAnalysisReport`

File: build.gradle.kts [hunk 3]
```diff
@@ -232,4 +235,58 @@ openApiMerger {
...
+            all { // all projects
+                onAny {
+                    severity(project.property("dependency.analysis").toString())
+                    exclude(
+                        "org.jetbrains:annotations",
+                        "com.fasterxml.jackson.datatype:jackson-datatype-jsr310",
+                        "com.fasterxml.jackson.core:jackson-core",
+                        "com.fasterxml.jackson.core:jackson-annotations",
+                        "com.fasterxml.jackson.core:jackson-databind",
+                        "com.fasterxml.jackson.datatype:jackson-datatype-jsr310",
```
Comment (algattik) [inline build.gradle.kts:259]: duplicate

File: build.gradle.kts [hunk 4]
```diff
@@ -232,4 +235,58 @@ openApiMerger {
...
+                onUsedTransitiveDependencies {
+                    severity("ignore")
+                }
+            }
+        }
+        abi {
+            exclusions {
+                excludeAnnotations(
+                        "io.opentelemetry.extension.annotations.WithSpan",
+                        "org.eclipse.dataspaceconnector.common.annotations.ExcludeFromDependecyAnalysisReport",
```
Comment (algattik) [inline build.gradle.kts:287]: typo "Dependecy"

File: buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
+package org.eclipse.dataspaceconnector.gradle;
+
+import org.gradle.api.Plugin;
+import org.gradle.api.Project;
+
+/*
```
Comment (algattik) [inline buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java:6]: move to top

File: buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+ *  terms of the Apache License, Version 2.0 which is available at
+ *  https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  SPDX-License-Identifier: Apache-2.0
+ *
+ *  Contributors:
+ *       Microsoft Corporation - initial API and implementation
+ *
+ */
+public class DependencyRulesPlugin implements Plugin<Project> {
```
Comment (algattik) [inline buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java:19]: add javadoc

File: extensions/iam/decentralized-identity/registration-service/build.gradle.kts [hunk 7]
```diff
@@ -7,13 +7,13 @@ val rsApi: String by project
...
-    implementation(project(":extensions:azure:events-config"))
-    implementation(project(":extensions:iam:decentralized-identity:identity-did-spi"))
+    api(project(":extensions:azure:events-config"))
+    api(project(":extensions:iam:decentralized-identity:identity-did-spi"))
 
     // third party
     implementation("com.squareup.okhttp3:okhttp:${okHttpVersion}")
     implementation("jakarta.ws.rs:jakarta.ws.rs-api:${rsApi}")
-    implementation("org.quartz-scheduler:quartz:2.3.0")
+    api("org.quartz-scheduler:quartz:2.3.0")
```
Comment (algattik) [inline extensions/iam/decentralized-identity/registration-service/build.gradle.kts:16]: false positive?

File: build.gradle.kts [hunk 8]
```diff
@@ -232,4 +235,58 @@ openApiMerger {
             }
         }
     }
-}
\ No newline at end of file
+}
+
+// Dependency analysis active if property "dependency.analysis" is set. Possible values are <'fail'|'warn'|'ignore'>.g po
```
Comment (Izzzu) [inline build.gradle.kts:240]: What's `.g po` in the end of the line?

File: extensions/iam/decentralized-identity/registration-service/build.gradle.kts [hunk 9]
```diff
@@ -7,13 +7,13 @@ val rsApi: String by project
 val okHttpVersion: String by project
 
 dependencies {
-    implementation(project(":extensions:azure:events-config"))
-    implementation(project(":extensions:iam:decentralized-identity:identity-did-spi"))
+    api(project(":extensions:azure:events-config"))
```
Comment (algattik) [inline extensions/iam/decentralized-identity/registration-service/build.gradle.kts:10]: false positive?

**Review by algattik — CHANGES_REQUESTED** [review]
Some dependencies promoted to api() look like false positive.

Consider reviewing them. Potentially reduce the PR to only change a couple of dep scopes for illustration, to simplify review.

File: build.gradle.kts (hunk 8 above)
Comment (marcgs) [reply build.gradle.kts:240]: "g po" is my shortcut for "git push" 😂 thanks for spotting this one

File: buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java (hunk 6 above)
Comment (marcgs) [reply buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesPlugin.java:23]: Done

**Comment by marcgs** [comment]
> Some dependencies promoted to api() look like false positive.
> 
> Consider reviewing them. Potentially reduce the PR to only change a couple of dep scopes for illustration, to simplify review.

Reduced changeset to 2 exemplary fixes

File: common/util/src/main/java/org/eclipse/dataspaceconnector/common/annotations/ExcludeFromDependencyAnalysisReport.java [hunk 10]
```diff
@@ -0,0 +1,30 @@
...
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Exclude a class from dependency analysis.
+ *
+ * @see <a href="https://github.com/autonomousapps/dependency-analysis-android-gradle-plugin">Gradle Dependency Analysis Plugin</a>
+ */
+@Retention(RetentionPolicy.RUNTIME)
+@Target({ElementType.METHOD, ElementType.TYPE})
```
Comment (Izzzu) [inline common/util/src/main/java/org/eclipse/dataspaceconnector/common/annotations/ExcludeFromDependencyAnalysisReport.java:28]: Does it make sense to have here also ElementType.METHOD?

**Review by Izzzu — APPROVED** [review]

File: build.gradle.kts [hunk 11]
```diff
@@ -18,6 +18,8 @@ plugins {
     checkstyle
     jacoco
     id("com.rameshkp.openapi-merger-gradle-plugin") version "1.0.4"
+    id ("org.eclipse.dataspaceconnector.dependency-rules") apply(false)
+    id("com.autonomousapps.dependency-analysis") version "1.0.0-rc04" apply (false)
```
Comment (algattik) [inline build.gradle.kts:22]: rc05 required for exclude annotations

File: buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+            });
+        }
+    }
+
+    private void dependencyError(String error) {
+        var message = format("DEPENDENCY RULE VIOLATION: %s", error);
+        if (getFailOnError().get()) {
+            throw new GradleException(message);
+        } else {
+            System.out.println(message);
```
Comment (algattik) [inline buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java:123]: is `System.out.println` this the recommended way to show output in a plugin?

File: build.gradle.kts [hunk 13]
```diff
@@ -232,4 +235,57 @@ openApiMerger {
...
+                onUsedTransitiveDependencies {
+                    severity("ignore")
+                }
+            }
+        }
+        abi {
+            exclusions {
+                excludeAnnotations(
+                        "io.opentelemetry.extension.annotations.WithSpan",
+                        "org.eclipse.dataspaceconnector.common.annotations.ExcludeFromDependecyAnalysisReport",
```
Comment (algattik) [inline build.gradle.kts:286]: typo is still present

**Review by algattik — CHANGES_REQUESTED** [review]
comments inline

File: build.gradle.kts (hunk 11 above)
Comment (marcgs) [reply build.gradle.kts]: rc04 is the versions required for working exclude annotations, but I will update the version anyways

File: build.gradle.kts [hunk 14]
```diff
@@ -232,4 +235,57 @@ openApiMerger {
...
+                }
+                onUsedTransitiveDependencies {
+                    severity("ignore")
+                }
+            }
+        }
+        abi {
+            exclusions {
+                excludeAnnotations(
+                        "io.opentelemetry.extension.annotations.WithSpan",
```
[the remark this answers is not in the document]
Comment (marcgs) [reply build.gradle.kts]: done

File: build.gradle.kts (hunk 13 above)
Comment (marcgs) [reply build.gradle.kts]: Removed the annotation

File: common/util/src/main/java/org/eclipse/dataspaceconnector/common/annotations/ExcludeFromDependencyAnalysisReport.java (hunk 10 above)
Comment (marcgs) [reply common/util/src/main/java/org/eclipse/dataspaceconnector/common/annotations/ExcludeFromDependencyAnalysisReport.java:28]: Annotation removed as it did not accomplish the goal is was meant for

File: buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java (hunk 12 above)
Comment (marcgs) [reply buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java:123]: Looking at samples in the documentation it seems that is the way indeed. This is analog to Kotlin's 'println' as well.

Comment (marcgs) [reply buildSrc/src/main/java/org/eclipse/dataspaceconnector/gradle/DependencyRulesTask.java:123]: I found that is possible to use a logger as well. Switching from `System.out` to the Task's logger.

**Review by ouphi — APPROVED** [review]

File: .github/workflows/verify.yaml [hunk 15]
```diff
@@ -26,12 +26,39 @@ jobs:
...
+        run: cat build/reports/dependency-analysis/build-health-report.txt
+
+  Unit-Test:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-java@v2
+        with:
+          distribution: 'temurin'
+          java-version: '11'
```
Comment (algattik) [inline .github/workflows/verify.yaml:52]: 17?

Comment (marcgs) [reply .github/workflows/verify.yaml]: done

**Pull request merged**